### PR TITLE
provider_core: Remove two TODO 3.0

### LINKED ISSUE #14377: Resolve or remove TODO 3.0 in crypto/provider_core.c
There are the following TODOs that need to be solved:
```
OSSL_LIB_CTX *ossl_provider_libctx(const OSSL_PROVIDER *prov)
{
    /* TODO(3.0) just: return prov->libctx; */
    return prov != NULL ? prov->libctx : NULL;
}
```
and
```
#ifndef FIPS_MODULE
/*
 * TODO(3.0) These error functions should use |handle| to select the proper
 * library context to report in the correct error stack, at least if error
 * stacks become tied to the library context.
 * We cannot currently do that since there's no support for it in the
 * ERR subsystem.
 */
static void core_new_error(const OSSL_CORE_HANDLE *handle)
{
    ERR_new();
}
```

### PR DESCRIPTION
**Pull request by t8m**
We need to keep the check for prov == NULL in ossl_provider_libctx
but it is not needed in core_get_libctx as there it can happen only when
there is a serious coding error in a third party provider and returning
NULL as libctx would be seriously wrong as that has a special meaning.

The second TODO is valid but not something that is relevant
for 3.0. Change it into a normal comment.

Fixes #14377

### CODE REVIEW

File: crypto/provider_core.c [hunk 1]
```diff
@@ -1099,7 +1098,13 @@ static OPENSSL_CORE_CTX *core_get_libctx(const OSSL_CORE_HANDLE *handle)
      */
     OSSL_PROVIDER *prov = (OSSL_PROVIDER *)handle;
 
-    return (OPENSSL_CORE_CTX *)ossl_provider_libctx(prov);
+    /*
```
Comment (slontis) [inline crypto/provider_core.c:1101]: Maybe assert here??

File: crypto/provider_core.c [hunk 2]
```diff
@@ -1121,11 +1126,9 @@ static int core_thread_start(const OSSL_CORE_HANDLE *handle,
  */
 #ifndef FIPS_MODULE
 /*
- * TODO(3.0) These error functions should use |handle| to select the proper
- * library context to report in the correct error stack, at least if error
+ * These error functions should use |handle| to select the proper
```
Comment (slontis) [inline crypto/provider_core.c:1129]: @mattcaswell would you confirm that this is ok?

File: crypto/provider_core.c [hunk 3]
```diff
@@ -1121,11 +1126,9 @@ static int core_thread_start(const OSSL_CORE_HANDLE *handle,
...
+ * library context to report in the correct error stack if error
  * stacks become tied to the library context.
- * We cannot currently do that since there's no support for it in the
- * ERR subsystem.
```
Comment (levitte) [inline crypto/provider_core.c:1128]: Why did you remove this part of the comment?  I see no reason to.

Comment (t8m) [reply crypto/provider_core.c:1128]: I thought it is kind of implied from the sentence before, but I'll add that back.

File: crypto/provider_core.c (hunk 1 above)
Comment (t8m) [reply crypto/provider_core.c:1101]: Added

**Comment by t8m** [comment]
Fixup pushed, please re-review.

File: crypto/provider_core.c [hunk 4]
```diff
@@ -1099,7 +1098,14 @@ static OPENSSL_CORE_CTX *core_get_libctx(const OSSL_CORE_HANDLE *handle)
      */
     OSSL_PROVIDER *prov = (OSSL_PROVIDER *)handle;
 
-    return (OPENSSL_CORE_CTX *)ossl_provider_libctx(prov);
+    /*
+     * Using ossl_provider_libctx would be wrong as that returns
+     * NULL for |prov| == NULL and NULL libctx has a special meaning
+     * that does not apply here. Here |prov| == NULL can happen only in
+     * case of coding error.
```
Comment (paulidale) [inline crypto/provider_core.c:1105]: case of _a_ coding error

**Review by paulidale — APPROVED** [review]

File: crypto/provider_core.c (hunk 4 above)
Comment (t8m) [reply crypto/provider_core.c]: fixup pushed, supposedly no reapproval is needed.

**Comment by t8m** [comment]
Merged to master. Thank you for the reviews. Unfortunately I did not notice the external-tests failure before merging, I am sorry for that. I've created #14571 to fix that.
